src/portalnet/protocol.rs

```diff
@@ -11,10 +11,19 @@ use super::{
 use super::{types::Message, Enr};
 use discv5::{Discv5ConfigBuilder, Discv5Event, TalkRequest};
 use log::{debug, error, warn};
-use tokio::sync::mpsc;
+use serde_json::Value;
+use tokio::sync::{mpsc, oneshot};
 
 use super::socket;
 
+type Responder<T, E> = oneshot::Sender<Result<T, E>>;
+
+#[derive(Debug)]
+pub enum PortalEndpoint {
+    NodeInfo { resp: Responder<Value, String> },
+    RoutingTableInfo { resp: Responder<Value, String> },
+}
```

Comment:
Something about this enum smells funny to me, but I can't put my finger on it. 🤷🏻‍♂️  carry on

Reply:
It could be factored a bit:

```rust
pub enum PortalEndpointKind {
    NodeInfo,
    RoutingTableInfo,
}

pub struct PortalEndpoint {
    kind: PortalEndpointKind,
    resp: Responder<Value, String>,
}
```